managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpScan.java

```diff
@@ -0,0 +1,124 @@
+    private final long timeOutMs;
+    private final long startTime = System.currentTimeMillis();
+
+    PositionImpl searchPosition;
+    Position lastMatchedPosition = null;
+
+    public OpScan(ManagedCursorImpl cursor, PositionImpl startPosition, Predicate<Entry> condition,
+                  ScanCallback callback, Object ctx, long maxEntries, long timeOutMs) {
+        this.cursor = cursor;
+        this.ledger = cursor.ledger;
+        this.startPosition = startPosition;
+        this.callback = callback;
+        this.condition = condition;
+        this.ctx = ctx;
+        this.searchPosition = startPosition;
+        this.remainingEntries.set(maxEntries);
+        this.timeOutMs = timeOutMs;
+    }
+
+
+    @Override
+    public void readEntryComplete(Entry entry, Object ctx) {
+        remainingEntries.decrementAndGet();
+        final Position position = entry.getPosition();
+        lastMatchedPosition = position;
+        // filter out the entry if it has been already deleted
+        // filterReadEntries will call entry.release if the entry is filtered out
+        List<Entry> entries = this.cursor.filterReadEntries(List.of(entry));
+        if (!entries.isEmpty()) {
+            boolean exit = false;
+            try {
+                if (!condition.apply(entry)) {
+                  exit = true;
+                }
+                entry.release();
+            } catch (Throwable err) {
+                log.error("[{}] user exception", cursor, err);
+                callback.scanFailed(ManagedLedgerException.getManagedLedgerException(err),
+                        Optional.ofNullable(searchPosition), OpScan.this.ctx);
+            }
+            if (exit) {
+                // user code requested to stop our scan
+                callback.scanComplete(lastMatchedPosition, ScanOutcome.USER_INTERRUPTED, OpScan.this.ctx);
+                return;
+            }
+        }
+        searchPosition = ledger.getPositionAfterN((PositionImpl) position, 1, PositionBound.startExcluded);
+        if (log.isDebugEnabled()) {
+            log.debug("readEntryComplete {} at {} next is {}", entry, position, searchPosition);
+        }
+
+        if (searchPosition.compareTo((PositionImpl) position) == 0) {
+            // we have reached the end of the ledger, as we are not doing progress
+            callback.scanComplete(lastMatchedPosition, ScanOutcome.COMPLETED, OpScan.this.ctx);
+            return;
+        }
+        find();
+    }
+
+    @Override
+    public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+        callback.scanFailed(exception, Optional.ofNullable(searchPosition), OpScan.this.ctx);
+    }
+
+    public void find() {
+        if (remainingEntries.get() <= 0) {
+            log.warn("[{}] Scan abort after reading too many entries", OpScan.this.cursor);
+            callback.scanComplete(lastMatchedPosition, ScanOutcome.ABORTED, OpScan.this.ctx);
```

Comment:
I am not sure it adds so much value.
If we add two distinct cases, then we have to add new enum values for new cases in the future.
The alternative would be to add a descriptive string, but usually it is not a good idea to provide human readable descriptions in APIs (no localization....)

This check is done only to prevent huge scans, the user will see "aborted: true", and that is enough to understand that the operation was interrupted abruptly. The logs will tell more.

This operation is intended to be run by administrators to understand the amount of messages that are still to be processed

Reply:
>This operation is intended to be run by administrators to understand the amount of messages that are still to be processed

Maybe `timestamp` is better? the broker searching entries until some specific time or timeout.